### PR TITLE
Enable git optimization

### PR DESCRIPTION
**Pull request by ppknap**
This PR enables `.git` optimization it also refactors prefetch logic and moves it to separate package. The optimization strategies are flexible so it will be possible to add new ones.

This logic also handles cases when localhost has no git executable but tries to mount remote git repository.

## Motivation and Context
Mount optimization.

## How Has This Been Tested?
Manually, Unit tests.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: go/src/koding/klient/machine/machinegroup/mount_test.go [hunk 1]
```diff
@@ -95,10 +95,11 @@ func TestAddMount(t *testing.T) {
 
 	// Add testing mount.
 	addMountReq := &AddMountRequest{
-		MountRequest{
+		MountRequest: MountRequest{
 			ID:    id,
 			Mount: m,
 		},
+		Strategies: []string{},
```
Comment (rjeczalik) [inline go/src/koding/klient/machine/machinegroup/mount_test.go]: Is there a need for this distinction between empty and nil slice?

File: go/src/koding/klient/machine/mount/sync/prefetch/git.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	return ".git/", int64(n.CountAll(-1)), n.DiskSizeAll(-1), nil
+}
+
+// PostRun runs git executable in order to restore downloaded state in working
+// directory.
+func (Git) PostRun(wd string) error {
+	var (
+		buf = bytes.Buffer{}
```
Comment (rjeczalik) [inline go/src/koding/klient/machine/mount/sync/prefetch/git.go]: style nit - just `buf bytes.Buffer`?

**Review by rjeczalik — APPROVED** [review]

File: go/src/koding/klient/machine/mount/sync/prefetch/git.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+)
+
+// Git prefetcher uses git to reduce the amount of prefetched data. The strategy
+// it uses is:
+//
+//  - download contnt of .git directory from remote.
+//  - run `git reset --hard`
+//  - any upstaged changes on remote will be detected and synced later by mount
+//    synchronization mechanisms.
+//
```
Comment (cihangir) [inline go/src/koding/klient/machine/mount/sync/prefetch/git.go]: * typo: `contnt`
* not clear what you mean by " any upstaged changes"

**Review by cihangir — APPROVED** [review]

File: go/src/koding/klient/machine/mount/sync/prefetch/git.go (hunk 3 above)
Comment (ppknap) [inline go/src/koding/klient/machine/mount/sync/prefetch/git.go]: another typo, it should have been `unstaged`. Will fix comments in next PR

**Pull request merged**